pkg/boot/universalpayload/utilities.go

```diff
@@ -685,8 +677,57 @@ func buildDeviceTreeInfo(buf io.Writer, mem *kexec.Memory, loadAddr uint64, rsdp
 		dt.NewNode("upl-custom", dt.WithProperty(
 			dt.PropertyU64("hoblistptr", loadAddr+tmpHobOffset),
 		)),
+	)), nil
+}
+
+func constructSMBIOS3Node() (*dt.Node, error) {
+	smbiosTableBase, size, err := getSMBIOSBase()
+
+	// According to EDK2 UPL implementation, only SMBIOS3 is supported in FDT.
+	if (err != nil) || (size != getSMBIOS3HdrSize()) {
+		return nil, errors.Join(ErrSMBIOS3NotFound, err)
+	}
+
+	return dt.NewNode("smbios", dt.WithProperty(
+		dt.PropertyString("compatible", "smbios"),
+		dt.PropertyRegion("reg", uint64(smbiosTableBase), uint64(pageSize)),
+	)), nil
+}
+
+func constructReservedMemoryNode(rsdpBase uint64) *dt.Node {
+	var rsvdNodes []*dt.Node
+
+	acpiChildNode := dt.NewNode("acpi", dt.WithProperty(
+		dt.PropertyString("compatible", "acpi"),
+		dt.PropertyRegion("reg", rsdpBase, uint64(pageSize)),
 	))
 
+	rsvdNodes = append(rsvdNodes, acpiChildNode)
+
+	smbios3ChildNode, err := constructSMBIOS3Node()
+	if err != nil {
+		// If we failed to retrieve SMBIOS3 information, prompt error
+		// message to indicate error message, and continue construct DTB.
+		fmt.Printf("WARNING: Failed to build SMBIOS3 Node (%v)\n", err)
```

Comment:
Is there some way we can avoid this Printf in a package? It causes real trouble when we are trying to read through tests, and it's generally not a good idea anyway. What might we do differently?

Reply:
An alternative way is to separate universalpayload.Load() method to universapayload.Load() and universayload.Exec(), which is quite similar to Linux kexec -l(--load) and kexec -e(--exec). Then we can print error message after  universapayload.Load()  returns and before execute universapayload.Exec()

Reply:
@rminnich  Hi Ron, how about new introduced commits, do they meet your requirements?

Reply:
@rminnich Hi Ron, would you mind help review whether new commits? Thanks.